autoload/lsp/ui/vim/completion.vim

```diff
@@ -0,0 +1,240 @@
+  if empty(l:user_data)
+    doautocmd User lsp_complete_done
+    return
+  endif
+
+  let s:context.line = getline('.')
+  let s:context.position = getpos('.')
+  let s:context.completed_item = copy(v:completed_item)
+  let s:context.server_name = l:user_data.server_name
+  let s:context.completion_item = l:user_data.completion_item
+  call feedkeys(printf("\<C-r>=<SNR>%d_on_complete_done_after()\<CR>", s:SID()), 'n')
+endfunction
+
+"
+" Apply textEdit or insertText(snippet) and additionalTextEdits.
+"
+function! s:on_complete_done_after() abort
+  let l:line = s:context.line
+  let l:position = s:context.position
+  let l:completed_item = s:context.completed_item
+  let l:server_name = s:context.server_name
+  let l:completion_item = s:context.completion_item
+
+  " check the commit characters are <BS> or <C-w>.
+  if strlen(getline('.')) < strlen(l:line)
+    doautocmd User lsp_complete_done
+    return ''
+  endif
+
+  " Do nothing if text_edit is disabled.
+  if !g:lsp_text_edit_enabled
+    doautocmd User lsp_complete_done
+    return ''
+  endif
+
+  let l:completion_item = s:resolve_completion_item(l:completion_item, l:server_name)
+
+  " apply textEdit or insertText(snippet).
+  let l:expand_text = s:get_expand_text(l:completed_item, l:completion_item)
+  if strlen(l:expand_text) > 0
+    call s:clear_inserted_text(
+          \   l:line,
+          \   l:position,
+          \   l:completed_item,
+          \   l:completion_item
+          \ )
+    if exists('g:lsp_snippets_expand_snippet') && len(g:lsp_snippets_expand_snippet) > 0
+      " vim-lsp-snippets expects commit characters removed.
+      call s:expand_text_simply(v:completed_item.word)
+    elseif exists('g:lsp_snippet_expand') && len(g:lsp_snippet_expand) > 0
+      " other snippet integartion point.
+      call g:lsp_snippet_expand[0](l:expand_text)
+    else
+      " expand text simply.
+      call s:expand_text_simply(l:expand_text)
+    endif
+  endif
+
+  " apply additionalTextEdits.
+  if has_key(l:completion_item, 'additionalTextEdits')
+    let l:saved_mark = getpos("'a")
+    let l:pos = getpos('.')
+    call setpos("'a", l:pos)
+    call lsp#utils#text_edit#apply_text_edits(
+          \ lsp#utils#get_buffer_uri(bufnr('%')),
+          \ l:completion_item.additionalTextEdits
+          \ )
+    let l:pos = getpos("'a")
+    call setpos("'a", l:saved_mark)
+    call setpos('.', l:pos)
+  endif
+
+  doautocmd User lsp_complete_done
+  return ''
+endfunction
+
+"
+" Try `completionItem/resolve` if it possible.
+"
+function! s:resolve_completion_item(completion_item, server_name) abort
+  " server_name is not provided.
+  if empty(a:server_name)
+    return a:completion_item
+  endif
+
+  " check server capabilities.
+  let l:capabilities = lsp#get_server_capabilities(a:server_name)
+  if !has_key(l:capabilities, 'completionProvider')
+        \ || !has_key(l:capabilities.completionProvider, 'resolveProvider')
```

Comment:
this can be set to `v:false` or `v:true`. Should we also check for the value?

Reply:
Ah, Thanks for your review.
It is my mistake. I've fixed it.